lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4193,23 +4317,98 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+      }
+
+      // 3. Build the adjoint successor cases for the `switch_enum` instruction.
+      //    The adjoint successors correspond to the predecessors of the current
+      //    block.
       SmallVector<std::pair<EnumElementDecl *, SILBasicBlock *>, 4>
-          predecessorCases;
+          adjointSuccessorCases;
       for (auto *predBB : bb->getPredecessorBlocks()) {
-        auto *adjointSuccBB = getAdjointBlock(predBB);
+        // Get the adjoint block and optional adjoint trampoline block of the
+        // predecessor block.
+        auto *adjointBB = getAdjointBlock(predBB);
+        auto *adjointTrampolineBB = getAdjointTrampolineBlock(predBB, bb);
+        SILBasicBlock *adjointSuccBB = nullptr;
+        // If the predecessor block does not have a corresponding adjoint
+        // trampoline block, then the adjoint successor is the adjoint block.
+        if (!adjointTrampolineBB) {
+          adjointSuccBB = adjointBB;
+        }
+        // Otherwise, the adjoint successor is the adjoint trampoline block,
+        // which branches to the adjoint block and propagates adjoint values of
+        // active values.
+        else {
+          adjointSuccBB = adjointTrampolineBB;
+          assert(adjointSuccBB && adjointSuccBB->getNumArguments() == 1);
+          SILBuilder adjointTrampolineBBBuilder(adjointSuccBB);
+          SmallVector<SILValue, 8> trampolineArguments;
+          // Propagate pullback struct argument.
+          trampolineArguments.push_back(adjointSuccBB->getArguments().front());
+          // Propagate adjoint values/buffers of active values/buffers.
+          auto &predBBActiveValues = activeValues[predBB];
+          for (auto activeValue : predBBActiveValues) {
+            if (activeValue->getType().isObject()) {
+              auto activeValueAdj = getAdjointValue(bb, activeValue);
+              auto concreteActiveValueAdj =
+                  materializeAdjointDirect(activeValueAdj, adjLoc);
+              trampolineArguments.push_back(concreteActiveValueAdj);
+              // If the adjoint block does not yet have a registered adjoint
+              // value for the active value, set the adjoint value to the
+              // forwarded adjoint value argument.
+              // TODO: Hoist this logic out of loop over predecessor blocks to
+              // remove the `hasAdjointValue` check.
+              if (!hasAdjointValue(predBB, activeValue)) {
+                auto *adjointBBArg =
+                    getActiveValueAdjointBlockArgument(predBB, activeValue);
+                auto forwardedArgAdj =
+                    makeConcreteAdjointValue(ValueWithCleanup(adjointBBArg));
+                initializeAdjointValue(predBB, activeValue, forwardedArgAdj);
+              }
+            } else {
+              // Propagate adjoint buffers using `copy_addr`.
+              auto adjBuf = getAdjointBuffer(bb, activeValue);
+              auto succAdjBuf = getAdjointBuffer(predBB, activeValue);
```

Comment:
The argument is `predBB`. Shouldn't the variable be called `predAdjBuf`?

Reply:
I think `predAdjBuf` may be a better name, but both names are somewhat confusing.

I originally named the variable `succAdjBuf` because it's a buffer in the *adjoint successor block*. But naming it as an *adjoint buffer of the predecessor block* seems less confusing since `predBB` is an argument, as you mentioned.

I guess the code here could be described as:
1. Propagating adjoint values/buffers to predecessor blocks.
2. Propagating adjoint values/buffers to adjoint successor blocks.

1 seems more natural to me, so I'll change to `predAdjBuf`.